README.md
# Traduttore Registry


[![Build Status](https://travis-ci.com/wearerequired/traduttore-registry.svg?branch=master)](https://travis-ci.com/wearerequired/traduttore-registry)
[![codecov](https://codecov.io/gh/wearerequired/traduttore-registry/branch/master/graph/badge.svg)](https://codecov.io/gh/wearerequired/traduttore-registry)
[![Latest Stable Version](https://poser.pugx.org/wearerequired/traduttore-registry/v/stable)](https://packagist.org/packages/wearerequired/traduttore-registry)
[![Latest Unstable Version](https://poser.pugx.org/wearerequired/traduttore-registry/v/unstable)](https://packagist.org/packages/wearerequired/traduttore-registry)

Allows loading translation files from a custom GlotPress site running [Traduttore](https://github.com/wearerequired/traduttore).

## Installation

If you're using [Composer](https://getcomposer.org/) to manage dependencies, you can use the following command to add the plugin to your site:

```bash
composer require wearerequired/traduttore-registry
```

After that, you can use `\Required\Traduttore_Registry\add_project( $type, $slug, $api_url )` in your theme or plugin.

*Parameters:*

* `$type`: either `plugin` or `theme`.
* `$slug`: must match the theme/plugin directory slug.
* `$api_url`: the URL to the Traduttore project translation API.

*Note:* On a multisite install it's recommended to use it in a must-use plugin.

## Usage

Here's an example of how you can use that function:

```php
\Required\Traduttore_Registry\add_project(
	'plugin',
	'example-plugin',
	'https://translate.example.com/api/translations/acme/acme-plugin/'
);

\Required\Traduttore_Registry\add_project(
	'theme',
	'example-theme',
	'https://translate.example.com/api/translations/acme/acme-theme/'
);
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.3] - 2020-09-12

### Changed
* Update `dealerdirect/phpcodesniffer-composer-installer` dev requirement from ^0.5.0 to ^0.7.0 for Composer 2 support.

## [2.1.2] - 2020-08-30

### Changed
* Exclude tests and config files from dist-archive. [#43]

## [2.1.1] - 2020-07-22

### Changed
* Call `wp_get_installed_translations()` and `get_available_languages()` only once to improve performance on plugins list table. [#39]
* Cache results of failed API requests to reduce retries on follow-up requests which may fail too. [#40]

## [2.1.0] - 2020-07-20

### Fixed
* Prevent PHP fatal error for malformed date strings in PO files. Props @nickcernis. [#32]

### Changed
* Only install updates for actually available languages. [#28]
* Performance: Make closures static where possible. [#30]
* Performance: Use fully qualified name for special compiled PHP functions. [#36]

## [2.0.0] - 2019-03-15

### Fixed
* Various documentation improvements.

### Changed
* Delete cached API results when plugins/themes are updated to avoid stale or missing translation updates.

### Added
* Integration tests.

### Removed
* Support for PHP 7.0.

## [1.0.2] - 2018-08-31

### Changed
* Improve support for new `wp language plugin|theme install|update` commands in WP-CLI 2.0.

## [1.0.1] - 2018-06-29

### Fixed
* Prevent a PHP notice that could occur under some circumstances.

## 1.0.0 - 2018-06-19

## Added
* Initial release

[Unreleased]: https://github.com/wearerequired/traduttore-registry/compare/2.1.3...HEAD
[2.1.3]: https://github.com/wearerequired/traduttore-registry/compare/2.1.2...2.1.3
[2.1.2]: https://github.com/wearerequired/traduttore-registry/compare/2.1.1...2.1.2
[2.1.1]: https://github.com/wearerequired/traduttore-registry/compare/2.1.0...2.1.1
[2.1.0]: https://github.com/wearerequired/traduttore-registry/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/wearerequired/traduttore-registry/compare/1.0.2...2.0.0
[1.0.2]: https://github.com/wearerequired/traduttore-registry/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/wearerequired/traduttore-registry/compare/1.0.0...1.0.1

[#28]: https://github.com/wearerequired/traduttore-registry/issues/28
[#30]: https://github.com/wearerequired/traduttore-registry/issues/30
[#32]: https://github.com/wearerequired/traduttore-registry/issues/32
[#36]: https://github.com/wearerequired/traduttore-registry/issues/36
[#39]: https://github.com/wearerequired/traduttore-registry/issues/39
[#40]: https://github.com/wearerequired/traduttore-registry/issues/40
[#43]: https://github.com/wearerequired/traduttore-registry/issues/43


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Required\Traduttore
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	echo "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL;
	exit( 1 );
}

$_root_dir = dirname( __DIR__, 2  );

if ( ! file_exists( dirname( __DIR__, 2  ). '/vendor/autoload.php' ) ) {
	echo "Could not find $_root_dir/vendor/autoload.php, have you run composer install ?" . PHP_EOL;
	exit( 1 );
}

putenv( "WP_TESTS_DIR=$_tests_dir" );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/tests/AddProject.php
<?php
/**
 * Class AddProject
 *
 * @package Traduttore\Tests
 */

namespace Required\Traduttore_Registry\Tests;

use DateTime;
use \WP_UnitTestCase;
use function \Required\Traduttore_Registry\add_project;

/**
 *  Test cases for the add_project() function.
 */
class AddProject extends WP_UnitTestCase {
	/**
	 * Verifies that calls to translation_api() are filtered.
	 */
	public function test_filter_translation_api(): void {
		require_once ABSPATH . 'wp-admin/includes/translation-install.php';

		$type     = 'plugin';
		$slug     = 'test-plugin';
		$api_url  = 'https://translate.required.com/api/translations/required/test-plugin/';
		$expected = [ 'foo' => 'bar' ];

		add_project( $type, $slug, $api_url );

		add_filter(
			'pre_http_request',
			static function ( $result, $args, $url ) use ( $api_url, $expected ) {
				if ( $api_url === $url ) {
					return [
						'headers'  => [],
						'body'     => json_encode( $expected ),
						'response' => [],
					];
				}

				return $result;
			},
			10,
			3
		);

		$actual = translations_api(
			'plugins',
			[
				'slug'    => $slug,
				'version' => '1.0',
			]
		);

		$this->assertSame( $expected, $actual );
	}

	/**
	 * Verifies that calls to translation_api() are filtered.
	 */
	public function test_do_not_filter_translation_api_for_other_projects(): void {
		require_once ABSPATH . 'wp-admin/includes/translation-install.php';

		$type     = 'plugin';
		$slug     = 'test-plugin';
		$api_url  = 'https://translate.required.com/api/translations/required/test-plugin/';
		$expected = [ 'foo' => 'bar' ];

		add_project( $type, $slug, $api_url );

		add_filter(
			'translations_api',
			static function ( $result ) use ( $expected ) {
				if ( false === $result ) {
					return $expected;
				}

				return $result;
			},
			100
		);

		$actual = translations_api(
			'plugins',
			[
				'slug'    => 'bar-plugin',
				'version' => '1.0',
			]
		);

		$this->assertSame( $expected, $actual );
	}

	/**
	 * Verifies that calls to wp_get_translation_updates() are filtered.
	 */
	public function test_get_translation_updates(): void {
		$type    = 'plugin';
		$slug    = 'internationalized-plugin'; // Part of the WordPress test suite.
		$api_url = 'https://translate.required.com/api/translations/required/internationalized-plugin/';
		$now     = ( new DateTime() )->format( 'Y-m-d H:i:s' );
		$locale  = 'de_DE';
		$package = 'https://translate.required.com/content/traduttore/internationalized-plugin-de_DE.zip';

		$body = [
			'translations' => [
				[
					'language' => $locale,
					'updated'  => $now,
					'package'  => $package,
				],
			],
		];

		$expected = [
			'language' => $locale,
			'updated'  => $now,
			'package'  => $package,
			'type'     => $type,
			'slug'     => $slug,
		];

		add_project( $type, $slug, $api_url );

		add_filter(
			'pre_http_request',
			static function ( $result, $args, $url ) use ( $api_url, $body ) {
				if ( $api_url === $url ) {
					return [
						'headers'  => [],
						'body'     => json_encode( $body ),
						'response' => [],
					];
				}

				return $result;
			},
			10,
			3
		);

		$actual = wp_get_translation_updates();

		$this->assertArraySubset( [ (object) $expected ], $actual );
	}

	/**
	 * Verifies that calls to wp_get_translation_updates() are not filtered
	 * when the remote package is older than the local one.
	 */
	public function test_get_translation_updates_outdated_package(): void {
		$type    = 'plugin';
		$slug    = 'internationalized-plugin'; // Part of the WordPress test suite.
		$api_url = 'https://translate.required.com/api/translations/required/internationalized-plugin/';
		$now     = ( new DateTime() )->setDate( 2000, 1, 1 )->format( 'Y-m-d H:i:s' );
		$locale  = 'de_DE';
		$package = 'https://translate.required.com/content/traduttore/internationalized-plugin-de_DE.zip';

		$body = [
			'translations' => [
				[
					'language' => $locale,
					'updated'  => $now,
					'package'  => $package,
				],
			],
		];

		add_project( $type, $slug, $api_url );

		add_filter(
			'pre_http_request',
			static function ( $result, $args, $url ) use ( $api_url, $body ) {
				if ( $api_url === $url ) {
					return [
						'headers'  => [],
						'body'     => json_encode( $body ),
						'response' => [],
					];
				}

				return $result;
			},
			10,
			3
		);

		$actual = wp_get_translation_updates();

		$this->assertEmpty( $actual );
	}

	/**
	 * Verifies that calls to wp_get_translation_updates() are not filtered
	 * when the API returns no translations.
	 */
	public function test_get_translation_updates_empty_api_result(): void {
		$type    = 'plugin';
		$slug    = 'internationalized-plugin'; // Part of the WordPress test suite.
		$api_url = 'https://translate.required.com/api/translations/required/internationalized-plugin/';
		$body    = [];

		add_project( $type, $slug, $api_url );

		add_filter(
			'pre_http_request',
			static function ( $result, $args, $url ) use ( $api_url, $body ) {
				if ( $api_url === $url ) {
					return [
						'headers'  => [],
						'body'     => json_encode( $body ),
						'response' => [],
					];
				}

				return $result;
			},
			10,
			3
		);

		$actual = wp_get_translation_updates();

		$this->assertEmpty( $actual );
	}

	/**
	 * Verifies that only translation updates for actually available languages are returned.
	 *
	 * @see https://github.com/wearerequired/traduttore-registry/issues/28
	 */
	public function test_get_translation_updates_available_languages(): void {
		$type    = 'plugin';
		$slug    = 'internationalized-plugin'; // Part of the WordPress test suite.
		$api_url = 'https://translate.required.com/api/translations/required/internationalized-plugin/';
		$now     = ( new DateTime() )->format( 'Y-m-d H:i:s' );
		$locale  = 'de_DE';
		$package = 'https://translate.required.com/content/traduttore/internationalized-plugin-de_DE.zip';

		$body = [
			'translations' => [
				[
					'language' => $locale,
					'updated'  => $now,
					'package'  => $package,
				],
				[
					'language' => 'de_CH',
					'updated'  => $now,
					'package'  => $package,
				],
			],
		];

		$expected = [
			'language' => $locale,
			'updated'  => $now,
			'package'  => $package,
			'type'     => $type,
			'slug'     => $slug,
		];

		add_project( $type, $slug, $api_url );

		add_filter(
			'pre_http_request',
			static function ( $result, $args, $url ) use ( $api_url, $body ) {
				if ( $api_url === $url ) {
					return [
						'headers'  => [],
						'body'     => json_encode( $body ),
						'response' => [],
					];
				}

				return $result;
			},
			10,
			3
		);

		$actual = wp_get_translation_updates();

		$this->assertArraySubset( [ (object) $expected ], $actual );
		$this->assertCount( 1, $actual );
	}
}


tests/phpunit/tests/SanitizeDate.php
<?php
/**
 * Class SanitizeDate
 *
 * @package Traduttore\Tests
 */

namespace Required\Traduttore_Registry\Tests;

use DateTime;
use \WP_UnitTestCase;
use function \Required\Traduttore_Registry\sanitize_date;

/**
 *  Tests dates are sanitized by the `sanitize_date()` function.
 */
class SanitizeDate extends WP_UnitTestCase {

	/**
	 * Ensures good dates are unaltered.
	 *
	 * @dataProvider good_dates
	 * @param string $date Unsanitized date string.
	 */
	public function test_good_dates_are_unaltered( $date ): void {
		$this->assertEquals( new DateTime( $date ), sanitize_date( $date ) );
	}

	/**
	 * Ensures timezones are stripped.
	 *
	 * @dataProvider dates_with_timezones
	 * @param string $date Unsanitized date string.
	 * @param string $expected Expected result after sanitization.
	 */
	public function test_timezones_are_stripped( $date, $expected ): void {
		$this->assertEquals( new DateTime( $expected ), sanitize_date( $date ) );
	}

	/**
	 * Ensures date strings DateTime cannot parse fall back to the Unix epoch.
	 *
	 * @dataProvider bad_dates
	 * @param string $date Unsanitized date string.
	 */
	public function test_bad_dates_give_unix_epoch( $date ): void {
		$epoch = new DateTime( '1970-01-01' );
		$this->assertEquals( $epoch, sanitize_date( $date ) );
	}

	/**
	 * Data provider for `test_good_dates_are_unaltered`.
	 *
	 * @return array Dates that should be unaltered after sanitization.
	 */
	function good_dates() {
		return [
			['2020-02-06 09:24:03+0000'],
			['2020-02-06 09:24:03'],
			['2020-02-06'],
		];
	}

	/**
	 * Data provider for `test_timezones_are_stripped`.
	 *
	 * @return array Unsanitized date and expected result pairs.
	 */
	function dates_with_timezones() {
		return [
			[
				'Tue Jan 12 2016 15:04:22 GMT+0100 (Romance Standard Time)',
				'Tue Jan 12 2016 15:04:22 GMT+0100',
			],
			[
				'Tue Dec 22 2015 12:52:19 GMT+0100 (West-Europa)',
				'Tue Dec 22 2015 12:52:19 GMT+0100',
			],
		];
	}

	/**
	 * Data provider for `test_bad_dates_give_unix_epoch`.
	 *
	 * @return array Date strings that DateTime will throw an Exception for.
	 */
	function bad_dates() {
		return [
			['TBC'],
			['123'],
			['-'],
			['The fifth of never.'],
		];
	}
}


inc/namespace.php
<?php
/**
 * Main Traduttore Registry library.
 *
 * @since 1.0.0
 *
 * Copyright (c) 2018-2020 required (email: info@required.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2 or, at
 * your discretion, any later version, as published by the Free
 * Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

namespace Required\Traduttore_Registry;

use DateTime;

const TRANSIENT_KEY_PLUGIN = 'traduttore-registry-plugins';
const TRANSIENT_KEY_THEME  = 'traduttore-registry-themes';

/**
 * Adds a new project to load translations for.
 *
 * @since 1.0.0
 *
 * @param string $type    Project type. Either plugin or theme.
 * @param string $slug    Project directory slug.
 * @param string $api_url Full GlotPress API URL for the project.
 */
function add_project( $type, $slug, $api_url ) {
	if ( ! has_action( 'init', __NAMESPACE__ . '\register_clean_translations_cache' ) ) {
		add_action( 'init', __NAMESPACE__ . '\register_clean_translations_cache', 9999 );
	}

	/**
	 * Short-circuits translations API requests for private projects.
	 */
	add_filter(
		'translations_api',
		static function ( $result, $requested_type, $args ) use ( $type, $slug, $api_url ) {
			if ( $type . 's' === $requested_type && $slug === $args['slug'] ) {
				return get_translations( $type, $args['slug'], $api_url );
			}

			return $result;
		},
		10,
		3
	);

	/**
	 * Filters the translations transients to include the private plugin or theme.
	 *
	 * @see wp_get_translation_updates()
	 */
	add_filter(
		'site_transient_update_' . $type . 's',
		static function ( $value ) use ( $type, $slug, $api_url ) {
			if ( ! $value ) {
				$value = new \stdClass();
			}

			if ( ! isset( $value->translations ) ) {
				$value->translations = [];
			}

			$translations = get_translations( $type, $slug, $api_url );

			if ( ! isset( $translations['translations'] ) ) {
				return $value;
			}

			$installed_translations = get_installed_translations( $type );
			$locales                = get_available_locales();

			/** This filter is documented in wp-includes/update.php */
			$locales        = apply_filters( $type . 's_update_check_locales', $locales );
			$active_locales = array_unique( $locales );

			foreach ( (array) $translations['translations'] as $translation ) {
				if ( ! \in_array( $translation['language'], $active_locales, true ) ) {
					continue;
				}

				if ( $translation['updated'] && isset( $installed_translations[ $slug ][ $translation['language'] ] ) ) {
					$local  = sanitize_date( $installed_translations[ $slug ][ $translation['language'] ]['PO-Revision-Date'] );
					$remote = new DateTime( $translation['updated'] );

					if ( $local >= $remote ) {
						continue;
					}
				}

				$translation['type'] = $type;
				$translation['slug'] = $slug;

				$value->translations[] = $translation;
			}

			return $value;
		}
	);
}

/**
 * Registers actions for clearing translation caches.
 *
 * @since 1.1.0
 */
function register_clean_translations_cache() {
	$clear_plugin_translations = static function() {
		clean_translations_cache( 'plugin' );
	};
	$clear_theme_translations  = static function() {
		clean_translations_cache( 'theme' );
	};

	add_action( 'set_site_transient_update_plugins', $clear_plugin_translations );
	add_action( 'delete_site_transient_update_plugins', $clear_plugin_translations );

	add_action( 'set_site_transient_update_themes', $clear_theme_translations );
	add_action( 'delete_site_transient_update_themes', $clear_theme_translations );
}

/**
 * Clears existing translation cache for a given type.
 *
 * @since 1.1.0
 *
 * @param string $type Project type. Either plugin or theme.
 */
function clean_translations_cache( $type ) {
	$transient_key = constant( __NAMESPACE__ . '\TRANSIENT_KEY_' . strtoupper( $type ) );
	$translations  = get_site_transient( $transient_key );

	if ( ! \is_object( $translations ) ) {
		return;
	}

	/*
	 * Don't delete the cache if the transient gets changed multiple times
	 * during a single request. Set cache lifetime to maximum 15 seconds.
	 */
	$cache_lifespan   = 15;
	$time_not_changed = isset( $translations->_last_checked ) && ( time() - $translations->_last_checked ) > $cache_lifespan;

	if ( ! $time_not_changed ) {
		return;
	}

	delete_site_transient( $transient_key );
}

/**
 * Gets the translations for a given project.
 *
 * @since 1.0.0
 *
 * @param string $type Project type. Either plugin or theme.
 * @param string $slug Project directory slug.
 * @param string $url  Full GlotPress API URL for the project.
 * @return array Translation data.
 */
function get_translations( $type, $slug, $url ) {
	$transient_key = constant( __NAMESPACE__ . '\TRANSIENT_KEY_' . strtoupper( $type ) );
	$translations  = get_site_transient( $transient_key );

	if ( ! \is_object( $translations ) ) {
		$translations = new \stdClass();
	}

	if ( isset( $translations->{$slug} ) && \is_array( $translations->{$slug} ) ) {
		return $translations->{$slug};
	}

	$result = json_decode( wp_remote_retrieve_body( wp_remote_get( $url, [ 'timeout' => 2 ] ) ), true );
	if ( ! \is_array( $result ) ) {
		// Cache an empty result in case of a failure
		// and retry on next update check.
		$result = [];
	}

	$translations->{$slug}       = $result;
	$translations->_last_checked = time();

	set_site_transient( $transient_key, $translations );

	return $result;
}

/**
 * Sanitizes a date string.
 *
 * DateTime fails to parse date strings that contain brackets, such as
 * “Tue Dec 22 2015 12:52:19 GMT+0100 (West-Europa)”, which appears in
 * PO-Revision-Date headers. Sanitization ensures such date headers are
 * parsed correctly into DateTime instances.
 *
 * @since 2.1.0
 *
 * @param string $date_string The date string to sanitize.
 * @return \DateTime Date from string if parsable, otherwise the Unix epoch.
 */
function sanitize_date( $date_string ) {
	$date_and_timezone = explode( '(', $date_string );
	$date_no_timezone  = trim( $date_and_timezone[0] );

	try {
		$date = new DateTime( $date_no_timezone );
	} catch ( \Exception $e ) {
		return new DateTime( '1970-01-01' );
	}

	return $date;
}

/**
 * Gets the installed translations.
 *
 * Results are cached.
 *
 * @since 2.2.0
 *
 * @see wp_get_installed_translations()
 *
 * @param string $type Project type. Either plugin or theme.
 * @return array Translation data.
 */
function get_installed_translations( $type ) {
	static $cache = [];

	if ( ! isset( $cache[ $type ] ) ) {
		$cache[ $type ] = wp_get_installed_translations( $type . 's' );
	}

	return $cache[ $type ];
}

/**
 * Gets all available and installed locales.
 *
 * Results are cached.
 *
 * @since 2.2.0
 *
 * @see get_available_languages()
 *
 * @return array List of installed locales.
 */
function get_available_locales() {
	static $cache = null;

	if ( ! isset( $cache ) ) {
		$cache = array_values( get_available_languages() );
	}

	return $cache;
}
